hail/src/main/scala/is/hail/expr/ir/IR.scala

```diff
@@ -235,6 +235,16 @@ final case class ArrayFlatMap(a: IR, name: String, body: IR) extends IR {
 }
 final case class ArrayFold(a: IR, zero: IR, accumName: String, valueName: String, body: IR) extends IR
 
+object ArrayFold2 {
+  def apply(a: ArrayFold): ArrayFold2 = {
```

Comment:
Would it be possible to make `ArrayFold.apply` call this and get rid of `ArrayFold` completely?

Reply:
this is part 2, but it's a larger refactor and wanted to isolate the new behavior in a single PR.